### PR TITLE
ARIA-236 Resumable workflow executions

### PR DESCRIPTION
**Pull request by mxmrlv**

### CODE REVIEW

**Comment by asfgit** [comment]
Can one of the admins verify this patch?

File: aria/cli/commands/executions.py [hunk 1]
```diff
@@ -134,18 +134,64 @@ def start(workflow_name,
...
+@aria.options.task_max_attempts()
+@aria.options.task_retry_interval()
+@aria.options.mark_pattern()
+@aria.options.verbose()
+@aria.pass_model_storage
+@aria.pass_resource_storage
+@aria.pass_plugin_manager
+@aria.pass_logger
+def resume(execution_id,
+           inputs,
```
Comment (mxmrlv) [inline aria/cli/commands/executions.py]: remove

File: aria/orchestrator/workflow_runner.py [hunk 2]
```diff
@@ -57,26 +57,29 @@ def __init__(self, workflow_name, service_id, inputs,
...
         # by several threads without raising errors on model objects shared between threads
-        self._service_id = service_id
 
-        self._validate_workflow_exists_for_service()
-
-        workflow_fn = self._get_workflow_fn()
-
-        execution = self._create_execution_model(inputs)
-        self._execution_id = execution.id
+        if workflow_name is not None and service_id is not None and execution_id is None:
```
Comment (mxmrlv) [inline aria/orchestrator/workflow_runner.py]: create single boolean

File: aria/orchestrator/workflow_runner.py [hunk 3]
```diff
@@ -57,26 +57,29 @@ def __init__(self, workflow_name, service_id, inputs,
...
+            self._service_id = service_id
+            self._workflow_name = workflow_name
+            self._validate_workflow_exists_for_service()
+            self._execution_id = self._create_execution_model(inputs).id
+        elif execution_id is not None:
+            self._execution_id = execution_id
+            self._service_id = self.execution.service.id
+            self._workflow_name = model_storage.execution.get(self._execution_id).workflow_name
+        else:
+            raise exceptions.InvalidWorkflowRunnerParams("")
```
Comment (mxmrlv) [inline aria/orchestrator/workflow_runner.py]: msg..

File: aria/orchestrator/workflow_runner.py [hunk 4]
```diff
@@ -86,12 +89,17 @@ def __init__(self, workflow_name, service_id, inputs,
...
+        if execution_id is None:
+            # Not an existing execution
+            workflow_fn = self._get_workflow_fn()
+            tasks_graph = workflow_fn(ctx=self._workflow_context, **execution_inputs_dict)
+            compile.create_execution_tasks(self._workflow_context, tasks_graph, executor.__class__)
 
         self._engine = engine.Engine(executors={executor.__class__: executor})
 
+        if workflow_name is None:
+            self._engine.resume_execution(self._workflow_context)
```
Comment (mxmrlv) [inline aria/orchestrator/workflow_runner.py]: move to function

File: aria/orchestrator/workflows/core/engine.py [hunk 5]
```diff
@@ -69,6 +69,10 @@ def execute(self, ctx):
             raise
 
     @staticmethod
+    def resume_execution(ctx):
+        events.on_resume_workflow_signal.send(ctx)
```
Comment (mxmrlv) [inline aria/orchestrator/workflows/core/engine.py]: ???

File: tests/orchestrator/test_workflow_runner.py [hunk 6]
```diff
@@ -293,3 +304,129 @@ def _create_workflow_runner(request, workflow_name, inputs=None, executor=None,
...
+def mock_workflow(ctx, graph):
+    node = ctx.model.node.get_by_name(tests_mock.models.DEPENDENCY_NODE_NAME)
+    graph.add_tasks(
+        api.task.OperationTask(
+            node, interface_name='aria.interfaces.lifecycle', operation_name='create')
+    )
+
+
+@operation
+def mock_success_task(**_):
```
Comment (mxmrlv) [inline tests/orchestrator/test_workflow_runner.py]: rename

File: tests/orchestrator/test_workflow_runner.py [hunk 7]
```diff
@@ -293,3 +304,129 @@ def _create_workflow_runner(request, workflow_name, inputs=None, executor=None,
...
+            plugin_manager=None,
+            workflow_name='custom_workflow',
+            executor=executor)
+        wf_thread = Thread(target=wf_runner.execute)
+        try:
+            wf_thread.start()
+
+            # Wait for the execution to start
+            while global_test_holder.get('state') != 'active':
+                pass
```
Comment (mxmrlv) [inline tests/orchestrator/test_workflow_runner.py]: sleep(0.1)

File: tests/orchestrator/test_workflow_runner.py [hunk 8]
```diff
@@ -293,3 +304,129 @@ def _create_workflow_runner(request, workflow_name, inputs=None, executor=None,
...
+            global_test_holder['state'] = 'terminated'
+            wf_runner.cancel()
+
+            # Make sure the execution was canceled and the task has not ended
+            while wf_runner.execution.status != workflow_context.execution.CANCELLED:
+                pass
```
Comment (mxmrlv) [inline tests/orchestrator/test_workflow_runner.py]: same

File: tests/orchestrator/test_workflow_runner.py [hunk 9]
```diff
@@ -293,3 +304,129 @@ def _create_workflow_runner(request, workflow_name, inputs=None, executor=None,
...
+            global_test_holder['resumed'] = True
+            new_wf_runner.execute()
+
+            # Wait for it to finish and assert changes.
+            while global_test_holder.get('state') != 'ended':
+                pass
+            assert task.status == task.SUCCESS
+            assert wf_runner.execution.status == wf_runner.execution.SUCCEEDED
+
+        except:
```
Comment (mxmrlv) [inline tests/orchestrator/test_workflow_runner.py]: no needed

File: tests/orchestrator/test_workflow_runner.py [hunk 10]
```diff
@@ -293,3 +304,129 @@ def _create_workflow_runner(request, workflow_name, inputs=None, executor=None,
...
+
+        wf_runner = WorkflowRunner(
+            service_id=workflow_context.service.id,
+            inputs={},
+            model_storage=workflow_context.model,
+            resource_storage=workflow_context.resource,
+            plugin_manager=None,
+            workflow_name='custom_workflow',
+            executor=executor)
+        wf_thread = Thread(target=wf_runner.execute)
```
Comment (mxmrlv) [inline tests/orchestrator/test_workflow_runner.py]: daemon

File: tests/orchestrator/test_workflow_runner.py [hunk 11]
```diff
@@ -293,3 +304,129 @@ def _create_workflow_runner(request, workflow_name, inputs=None, executor=None,
...
+        api.task.OperationTask(
+            node, interface_name='aria.interfaces.lifecycle', operation_name='create')
+    )
+
+
+@operation
+def mock_success_task(**_):
+    global_test_holder['state'] = 'active'
+    while global_test_holder.get('state') != 'terminated':
+        if global_test_holder.get('resumed') is True:
```
Comment (mxmrlv) [inline tests/orchestrator/test_workflow_runner.py]: wait

File: tests/orchestrator/test_workflow_runner.py [hunk 12]
```diff
@@ -327,47 +335,39 @@ def test_resume_workflow(self, workflow_context, executor):
...
-
-        except:
-            global_test_holder['state'] = 'terminated'
-            wf_thread.join(0.5)
-            raise
+        wf_thread.daemon = True
+        wf_thread.start()
+
+        # Wait for the execution to start
+        events['is_active'].wait(5)
```
Comment (mxmrlv) [inline tests/orchestrator/test_workflow_runner.py]: raise?

**Pull request merged**